File: HwangSettong/week_3/168p.py
# https://school.programmers.co.kr/learn/courses/30/lessons/64061
import numpy as np
def solution(board, moves):
    stack = []
    answer = 0
    board = np.array(board).T.tolist()
    for i in range(0,len(board)):
        board[i].reverse()
        board[i] = board[i][0:board[i].index(0)]
    for i in moves:
        if len(board[i-1]) == 0:
            continue
        temp = board[i-1][-1]
        board[i-1] = board[i-1][:-1]
        if stack[-1:] == [temp]:
            stack = stack[:-1]
            answer += 2
        else :
            stack.append(temp)
    return answer

File: HwangSettong/week_3/168p.java
// https://school.programmers.co.kr/learn/courses/30/lessons/64061
import java.util.*;
class Solution {
    public int solution(int[][] board, int[] moves) {
        HashMap<Integer,Stack<Integer>> stacks = new HashMap<Integer,Stack<Integer>>();
        Stack<Integer> stack;
        for(int key=0; key<board.length; key++){
            stack = new Stack<Integer>();
            for(int i = board.length - 1; i >= 0 ; i--){
                if(board[i][key] == 0){break;}
                stack.push(board[i][key]);
            }
            stacks.put(key+1, stack);
        }
        
        stack = new Stack<Integer>();
        int answer = 0, temp;
        for(int m : moves){
            if(stacks.get(m).isEmpty()){
                continue;
            }
            temp = stacks.get(m).pop();
            if(!stack.isEmpty() && stack.peek() == temp){
                stack.pop();
                answer+=2;
            }else{
                stack.push(temp);
            }
        }
        return answer;
    }
}
